File: leetcode0350/solution.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        /* 
        Time: O(N log N + M log M) | Space: O(N + M)

        N, M: Length of nums1, nums2

        Explanation:
        The reason why there are log in Time Complexity
        is because map sorts the keys in ascending order automatically.
        */
        map<int, int> count;

        for (int& num: nums1) {
            if (count.find(num) == count.end()) {
                count[num] = 1;
            } else {
                count[num] += 1;
            }
        }

        vector<int> res;

        for (int& num: nums2) {
            if (count.find(num) == count.end()) {
                continue;
            } else {
                if (count[num] > 0) {
                    count[num] -= 1;
                    res.push_back(num);
                }
            }
        }

        return res;
    }
};

File: leetcode0350/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn intersect(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        /// Time: O(n + m) | Space: O(n + m)
        /// 
        /// where n is the length of nums1
        /// and m is the length of nums2
        let mut map: HashMap<i32, i32> = HashMap::new();

        for num in nums1 {
            map.entry(num).and_modify(|cnt| *cnt += 1).or_insert(1);
        }

        let mut res: Vec<i32> = Vec::new();

        for num in nums2 {
            map.entry(num).and_modify(|cnt| {
                if *cnt > 0 {
                    res.push(num);
                    *cnt -= 1;
                }
            });
        }

        res
    }
}